In [2]:
import pandas as pd 

In [3]:
# Upload the 2023 FEVS data
df_2023 = pd.read_csv('/Users/sophie/Desktop/ADS599_CapstoneProject/data/FEVS2023_PRDF_CSV/FEVS_2023_PRDF.csv')
df_2023.head(10)

RandomID agency   Q1   Q2   Q3   Q4   Q5   Q6   Q7 Q8  ...  DHISP  \
0  112970976817     XX  4.0  4.0  4.0  4.0  2.0  3.0  5.0  4  ...      B   
1  194868625278     XX  5.0  4.0  5.0  5.0  3.0  4.0  4.0  3  ...      B   
2  152966380283     XX  3.0  2.0  4.0  4.0  4.0  4.0  4.0  2  ...      B   
3  193041162980     XX  5.0  4.0  5.0  5.0  5.0  4.0  5.0  5  ...      B   
4  146655962451     XX  3.0  3.0  3.0  3.0  4.0  2.0  4.0  4  ...    NaN   
5  195312241136     XX  4.0  5.0  4.0  5.0  5.0  5.0  5.0  5  ...      B   
6  187815423862     XX  5.0  5.0  5.0  5.0  4.0  4.0  5.0  4  ...      B   
7  174698178690     XX  2.0  4.0  4.0  2.0  3.0  NaN  4.0  2  ...      B   
8  190481568495     XX  4.0  3.0  NaN  4.0  2.0  4.0  4.0  4  ...      B   
9  138139604049     XX  5.0  5.0  5.0  5.0  5.0  5.0  5.0  5  ...      A   

   DDIS  DAGEGRP  DSUPER  DFEDTEN  DSEX  DMIL  DLEAVING    POSTWT  Year  
0     B        B       B        B     A     B         C  3.007383  2023  
1     B        B       B        C     A     A         A  2.317204  2023  
2     B        B       A        A     A     A         B  2.119756  2023  
3     A        A       A        A     B     A         A  2.007827  2023  
4   NaN      NaN     NaN      NaN   NaN   NaN       NaN  2.111567  2023  
5     B        B       B        C     A     A         B  2.151488  2023  
6     B        B       A        B     B     B         C  2.411118  2023  
7     B        A       B        A     B     B         C  2.007827  2023  
8     B        B       B        C     A     A         B  2.180044  2023  
9     A        B       B        A     A     A         A  1.923460  2023  

[10 rows x 109 columns]

In [3]:
df_2023.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 625568 entries, 0 to 625567
Columns: 109 entries, RandomID to Year
dtypes: float64(44), int64(2), object(63)
memory usage: 520.2+ MB


In [5]:
df_2023.describe()

RandomID             Q1             Q2             Q3  \
count  6.255680e+05  621922.000000  616048.000000  616102.000000   
mean   1.500161e+11       3.833506       3.726676       3.919036   
std    2.887526e+10       1.086167       1.174349       1.066453   
min    1.000002e+11       1.000000       1.000000       1.000000   
25%    1.250110e+11       3.000000       3.000000       3.000000   
50%    1.499730e+11       4.000000       4.000000       4.000000   
75%    1.750317e+11       5.000000       5.000000       5.000000   
max    1.999999e+11       5.000000       5.000000       5.000000   

                  Q4             Q5             Q6             Q7  \
count  615429.000000  620833.000000  614430.000000  620335.000000   
mean        4.076792       3.490584       3.647810       4.182343   
std         0.943897       1.223422       1.167928       0.883019   
min         1.000000       1.000000       1.000000       1.000000   
25%         4.000000       3.000000       3.000000       4.000000   
50%         4.000000       4.000000       4.000000       4.000000   
75%         5.000000       4.000000       4.000000       5.000000   
max         5.000000       5.000000       5.000000       5.000000   

                  Q9            Q10  ...            Q71            Q72  \
count  623096.000000  617957.000000  ...  600258.000000  600439.000000   
mean        3.828930       3.711669  ...       3.471209       3.658508   
std         0.983467       1.069211  ...       1.215841       1.118704   
min         1.000000       1.000000  ...       1.000000       1.000000   
25%         3.000000       3.000000  ...       3.000000       3.000000   
50%         4.000000       4.000000  ...       4.000000       4.000000   
75%         4.000000       4.000000  ...       4.000000       4.000000   
max         5.000000       5.000000  ...       5.000000       5.000000   

                 Q86            Q87            Q88            Q89  \
count  596094.000000  595301.000000  597272.000000  595989.000000   
mean        3.685904       3.944027       3.692417       4.072010   
std         1.120587       1.040765       1.149891       0.924539   
min         1.000000       1.000000       1.000000       1.000000   
25%         3.000000       4.000000       3.000000       4.000000   
50%         4.000000       4.000000       4.000000       4.000000   
75%         5.000000       5.000000       5.000000       5.000000   
max         5.000000       5.000000       5.000000       5.000000   

                 Q90            Q91         POSTWT      Year  
count  597599.000000  597994.000000  625568.000000  625568.0  
mean        4.421753       1.810241       2.703555    2023.0  
std         0.718301       0.887186       1.938572       0.0  
min         1.000000       1.000000       0.614624    2023.0  
25%         4.000000       1.000000       1.521207    2023.0  
50%         5.000000       2.000000       1.975669    2023.0  
75%         5.000000       2.000000       3.101316    2023.0  
max         5.000000       4.000000      25.298883    2023.0  

[8 rows x 46 columns]

In [22]:
# Categorize the survey data based on the codebook
categories = {
    "Demographics": ['DHISP', 'DDIS', 'DAGEGRP', 'DSEX', 'DSUPER', 'DFEDTEN', 'DMIL', 'DLEAVING'],
    "Golbal_Satisfaction": ['Q46', 'Q70', 'Q71', 'Q72'],
    "Performance_Confidence": ['Q20', 'Q21', 'Q22', 'Q23'],
    "Employee_Engagement": ['Q2', 'Q3', 'Q4', 'Q6', 'Q7', 'Q58', 'Q59', 'Q61', 'Q62', 'Q48', 'Q50', 'Q51', 'Q52', 'Q54'],
    "DEIA": ['Q73', 'Q74', 'Q75', 'Q76', 'Q77', 'Q78', 'Q79', 'Q80', 'Q81', 'Q82', 'Q83', 'Q84', 'Q85'],
    "Employee_Experience": ['Q86', 'Q87', 'Q88', 'Q89', 'Q90'],
    "Performance_Dimensions": {
        "Agile": ['Q13', 'Q14', 'Q27', 'Q29', 'Q30', 'Q64', 'Q28', 'Q31', 'Q42', 'Q65'],
        "Employee-Focused": ['Q1', 'Q10', 'Q25', 'Q66', 'Q67', 'Q36', 'Q43', 'Q44', 'Q5', 'Q34', 'Q49', 'Q63'],
        "Foundations": ['Q40', 'Q60', 'Q68', 'Q15', 'Q18', 'Q32', 'Q33', 'Q8', 'Q45', 'Q9', 'Q19', 'Q24'],
        "Goal_Oriented": ['Q11', 'Q53', 'Q26', 'Q38', 'Q12', 'Q55', 'Q56', 'Q17', 'Q35', 'Q69'],
        "Other": ['Q37', 'Q39', 'Q41', 'Q47']
    }
    
}

# Create new order 
new_order = (
    categories['Demographics'] +
    categories['Golbal_Satisfaction'] +
    categories['Performance_Confidence'] +
    categories['Employee_Engagement'] +
    categories['DEIA'] +
    categories['Employee_Experience'] +
    categories['Performance_Dimensions']['Agile'] +
    categories['Performance_Dimensions']['Employee-Focused'] +
    categories['Performance_Dimensions']['Foundations'] +
    categories['Performance_Dimensions']['Goal_Oriented'] +
    categories['Performance_Dimensions']['Other'] +
     ['POSTWT', 'Year']
)

df = df_2023[new_order]

In [26]:
# Clean the data 
df = df.dropna(how='all')  # Drop rows where all elements are NaN
df = df.dropna(axis=1, how='all')  # Drop columns where all elements are NaN

# Map Likert scale values to numerical values
likert_map = {
    'A': 5,
    'B': 4,
    'C': 3,
    'D': 2,
    'E': 1,
    'X': None
}
# exclude last 2 columns
df = df.iloc[:, :-2].replace(likert_map).apply(pd.to_numeric, errors='coerce') # Exclude the last two columns (POSTWT, Year)



/var/folders/j8/2g7hyn014h73tt7fchfsv7zc0000gn/T/ipykernel_19201/1615806625.py:14: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.iloc[:, :-2].replace(likert_map).apply(pd.to_numeric, errors='coerce') # Exclude the last two columns (POSTWT, Year)


In [27]:
df.head()

DHISP  DDIS  DAGEGRP  DSEX  DSUPER  DFEDTEN  DMIL  DLEAVING  Q46  Q70  ...  \
0    4.0   4.0      4.0   5.0     4.0      4.0   4.0       3.0  4.0  4.0  ...   
1    4.0   4.0      4.0   5.0     4.0      3.0   5.0       5.0  4.0  4.0  ...   
2    4.0   4.0      4.0   5.0     5.0      5.0   5.0       4.0  2.0  2.0  ...   
3    4.0   5.0      5.0   4.0     5.0      5.0   5.0       5.0  5.0  5.0  ...   
4    NaN   NaN      NaN   NaN     NaN      NaN   NaN       NaN  4.0  3.0  ...   

   Q12  Q55  Q56  Q17  Q35  Q69  Q37  Q39  Q41  Q47  
0  NaN  2.0  1.0  4.0  4.0  3.0  4.0  4.0  4.0  2.0  
1  4.0  3.0  3.0  3.0  3.0  3.0  4.0  4.0  4.0  3.0  
2  4.0  4.0  4.0  NaN  2.0  3.0  4.0  2.0  2.0  1.0  
3  4.0  5.0  5.0  4.0  4.0  4.0  5.0  NaN  3.0  4.0  
4  2.0  3.0  4.0  2.0  4.0  4.0  5.0  2.0  2.0  2.0  

[5 rows x 96 columns]

In [ ]:
df.isna().sum()  # Check for missing values

DHISP      71064
DDIS       64517
DAGEGRP    63572
DSEX       65663
DSUPER     41053
           ...  
Q69        26213
Q37        22965
Q39        54913
Q41        36390
Q47        50284
Length: 96, dtype: int64

In [37]:
# Drop all rows with NaN values
df_clean = df.dropna()

#df_clean.isna().sum()  # Verify that there are no NaN values left

df_clean.shape

(141096, 96)

Since the dataset has over 625,000 rows, removing rows with missing values is a simple strategy leaving us with 141,096 rows to analyze.

